### DISCUSSION TITLE
If ems.sys isn't loaded, dpmi apps are silently terminated

### DISCUSSION THREAD

**Discussion by Baron-von-Riedesel** [dosemu2/dosemu2#1722] [General]
Well, you get 2 error msgs on the console:

ERROR: MSDOS: EMS is disabled
ERROR: RM seg not set

and the log shows that internally dosemu calls real-mode int 0x67 from protected-mode, but why?
At least there should be an error msg at the DOS console that the client has been terminated!

[boot.log.zip](https://github.com/dosemu2/dosemu2/files/9051382/boot.log.zip)

**Comment by stsp** [comment]
Because it uses ems for translation
buffer?
I dont think there is an explicit termination,
just an msdos call fails I think.

**Comment by Baron-von-Riedesel** [comment]
> just an msdos call fails I think.

That's correct - the problem is: most apps are not capable to handle this condition properly.
The HX DPMI loaders call the vendor API "MS-DOS" to check if a dos extender is available.
Perhaps DosEmu should make this call fail ( return with al != 0 ) if EMS.SYS isn't loaded?

Also, it would be good if there's a possibility to suppress those EMS calls in the log - they increase the already bloated log considerably.

**Reply by stsp** [reply]
> Perhaps DosEmu should make this call fail ( return with al != 0 ) if EMS.SYS isn't loaded?

Problematic.
- Any emm can be loaded at any
  time, including emm286.exe.
- Most msdos calls do not require
  emm buffer. They use the small,
  client-allocated area, and only
  the large ones (like I/O) would
  fail w/o emm.
- I simply don't understand why
  everyone is trying to disable ems.
dosemu2 gives you an unprecedented
amount of free UMB space. If people
will keep doing that (and you are not
the first one, of course), I'll simply
permanently allocate an UMB for
msdos calls. This will finally allow
to disable emm and everyone will
be very happy! No one will even
notice that the UMBs have actually
reduced, not enlarged. I am quite
sure those who disable emm, do
not even check the amount of UMBs,
they are absolutely ignorant. :)
But those who actually care about
an UMB space, would suffer...

> Also, it would be good if there's a possibility to suppress those EMS calls in the log

This is more than welcome.
Please put any "annoying" logging
under `if (debug_level('M') >= 8)`
or so, and submit a patch.

**Reply by stsp** [reply]
Ah, although I can mark the
168ah call as requiring the
buffer. In this case it will fail
w/o emm, helping your detection.

**Comment by stsp** [comment]
So I applied a "fix".
But I am not sure HX can
work around missing MS-DOS
extension. I have a pmdapi
project where I tried to provide
the MS-DOS extension by
DPMI client, and that appears
to be very difficult.

Also I was wrong that just
the msdos call fails: no, client
is explicitly terminated.
I avoided that for 168ah now,
allowing it to fail.

**Comment by Baron-von-Riedesel** [comment]
> I simply don't understand why everyone is trying to disable ems.
> dosemu2 gives you an unprecedented amount of free UMB space. If people ...

:))

I didn't care about UMBs when opening the discussion.
I'm currently analyzing what's needed to make OpenWatcom's debugger WD run in DosEmu and debug other protected-mode applications - locally.
The "standard" WD runs already, but to debug dpmi apps special requirements must be met which currently aren't - and hence WD crashes in DosEmu.
To get a better overview of what's happening I realized that those EMS calls are spoiling the log - and wondered it they could be removed; my first step was to disable loading ems.sys ...

**Reply by stsp** [reply]
> my first step was to disable loading ems.sys ...

Ok, cool. :)
So would you like, as a second
step, just put them under a
proper debug_level() check?

**Comment by andrewbird** [comment]
> no, client
is explicitly terminated.

I had assumed that it wasn't possible for dosemu to terminate a DOS process. Is that something you could do in the case of DPMI host stack exhaustion #1687 instead of exiting dosemu?

**Comment by Baron-von-Riedesel** [comment]
> So would you like ... just put them under a proper debug_level() check?

I'd love to do that, but the DosEmu source is soo complex, that's simply beyond my skills ....

**Reply by hex-4d5a9000** [reply]
"Beyond my skills"? Seriously? You made a DOS Extender ENTIRELY in x86 assembly, didn't you?

**Comment by stsp** [comment]
Well, you definitely can find the proper
logging strings and put them under if...
just try! :)

**Comment by stsp** [comment]
Yes, I can terminate dpmi prog while
its in protected mode (not in real mode),
but thats not reliable.
Havent you seen the message (under
eg qdpmi) "the app was terminated.
your system is likely unstable and may
need to be rebooted".
So I just dont care much, and in case
of some stack exhaustion or corruption,
the clean shutdown is much less likely
than in case of some msdos call.

**Reply by andrewbird** [reply]
okay, fair enough.
